src/sympc/protocol/falcon/falcon.py

```diff
@@ -73,93 +77,293 @@ def __eq__(self, other: Any) -> bool:
 
+            # Convert 3-3 shares to 2-3 shares by resharing
+            reshared_shares = ReplicatedSharedTensor.distribute_shares(
+                z_shares, x.session
+            )
+            result = MPCTensor(shares=reshared_shares, session=x.session)
+        result.shape = MPCTensor._get_shape(op_str, x.shape, y.shape)  # for prrs
         return result
 
+    @staticmethod
+    def triple_verification(
+        z_sh: ReplicatedSharedTensor,
+        eps: torch.Tensor,
+        delta: torch.Tensor,
+        op_str: str,
+        **kwargs,
+    ) -> ReplicatedSharedTensor:
+        """Performs Beaver's triple verification check.
+
+        Args:
+            z_sh (ReplicatedSharedTensor) : share of multiplied value(x*y).
+            eps (torch.Tensor) :masked value of x
+            delta (torch.Tensor): masked value of y
+            op_str (str): Operator string.
+            kwargs: Keywords arguments for the operator.
+
+        Returns:
+            ReplicatedSharedTensor : Result of the verification.
+        """
+        session = get_session(z_sh.session_uuid)
+
+        crypto_store = session.crypto_store
+        eps_shape = tuple(eps.shape)
+        delta_shape = tuple(delta.shape)
+
+        primitives = crypto_store.get_primitives_from_store(
+            f"beaver_{op_str}", eps_shape, delta_shape
+        )
+
+        a_share, b_share, c_share = primitives
+
+        op = getattr(operator, op_str)
+
+        eps_delta = op(eps, delta, **kwargs)
+        eps_b = b_share.clone()
+        delta_a = a_share.clone()
+
+        # prevent re-encoding as the values are encoded.
+        # TODO: should be improved.
+        for i in range(2):
+            eps_b.shares[i] = op(eps, eps_b.shares[i])
+            delta_a.shares[i] = op(delta_a.shares[i], delta)
+
+        rst_share = c_share + delta_a + eps_b
+
+        if session.rank == 0:
+            rst_share.shares[0] = rst_share.shares[0] + eps_delta
+
+        if session.rank == 2:
+            rst_share.shares[1] = rst_share.shares[1] + eps_delta
+
+        return rst_share
+
+    @staticmethod
+    def falcon_mask(
+        x_sh: ReplicatedSharedTensor, y_sh: ReplicatedSharedTensor, op_str: str
+    ) -> Tuple[ReplicatedSharedTensor, ReplicatedSharedTensor]:
+        """Falcon mask.
+
+        Args:
+            x_sh (ReplicatedSharedTensor): X share
+            y_sh (ReplicatedSharedTensor) : Y share
+            op_str (str): Operator
+
+        Returns:
+            Tuple[ReplicatedSharedTensor,ReplicatedSharedTensor] : Masked values.
+        """
+        session = get_session(x_sh.session_uuid)
+
+        crypto_store = session.crypto_store
+
+        primitives = crypto_store.get_primitives_from_store(
+            f"beaver_{op_str}", x_sh.shape, y_sh.shape, remove=False
+        )
+
+        a_sh, b_sh, _ = primitives
+
+        return x_sh - a_sh, y_sh - b_sh
+
+    @staticmethod
+    def mul_malicious(
```

Comment:
Q: The implementation does not look like what they say in the paper? (in the paper, for malicious setup, they say that x_i is sent to P_i+1 and x_i+1 to P_i-1,

Reply:
That's for reconstruction. Here the implementation of malicious mult is from ABY3. The reason why the reconstruction verification does not work is that parties have only a single share after computing Z, reconstruction verification requires 2 per party. So in order to still provide malicious security, ABY3 paper (actually Lindell but ABY3 extends it to arithmetic), there is this protocol where they compute beaver triplets on original shares used for mult.

Reply:
It's explained in ABY3 paper section 9.1 or 9.2 clearly.

Reply:
Adding on to what Hrishi said, reconstruction detects malicious behavior ,but it also reveals the underlying secret, we do not want to do this , in the middle of the computation(leaking intermediate values unless we intend to),for verification we do the triple check, george you think of it similar like spdz , where you use beaver triples for multiplication itself, here we have a custom protocol for multiplication , but use the triple as a verifier ,to veriy that the above multiplication is correct.

Reply:
@kamathhrishi yep - my bad...checking it now with the ABY3 details